### ISSUE TITLE
what is the variable '_dbt_max_partition' for? It looks unuseful, costs a lot of processed bytes.

### ISSUE THREAD

**Issue by xufengtian14** [dbt-labs/dbt#3666]
https://github.com/dbt-labs/dbt/blob/72c831a80a9439a4ec7e472472cf6451c88f0223/plugins/bigquery/dbt/include/bigquery/macros/materializations/incremental.sql#L52

**Comment by xufengtian14** [comment]
If a partitioned table is enough big, the SQL '''select max({{ partition_by.field }}) from {{ this }}''' still costs a lot of processed bytes.

**Comment by jtcohen6** [comment]
Hey @xufengtian14, totally fair! We initially added `_dbt_max_partition` in v0.16 as a "nice-to-have" scripting variable, that demonstrated the value BigQuery scripts. At the time, we were surprised at the lack of costless options for getting max partition values in standard SQL, since there are free metadata-only queries that can be run in legacy SQL.

Unfortunately, in the 1.5 years since then, the BigQuery team has not made moves toward making `select max(partition_col) from table` a metadata-only (free) query. So I'd be in favor of offering an option to disable `_dbt_max_partition`, perhaps by default, as well as other options to achieve the same result.

I'm going to close this issue in favor of https://github.com/dbt-labs/dbt/issues/2278, which has slightly more detail about the direction we want to see this go. That one is tagged a "good first issue," meaning that I'd be thrilled to support a community member who'd be interested in taking it on as a contribution! Is that something you might be interested in giving a go?

**Comment by bcolbert978** [comment]
Hey @jtcohen6 - I know this is closed but feels more relevant here than in #2278 - re: your comment above:
> Unfortunately, in the 1.5 years since then, the BigQuery team has not made moves toward making `select max(partition_col) from table` a metadata-only (free) query. So I'd be in favor of offering an option to disable `_dbt_max_partition`, perhaps by default, as well as other options to achieve the same result.

Although still not technically free (10MB min per query) and still in preview, it looks like the [INFORMATION_SCHEMA.PARTITIONS](https://cloud.google.com/bigquery/docs/information-schema-tables#partitions_view) does get us closer to a spot where including `_dbt_max_partition` by default is not onerous. The column scan at the outset of an incremental update script, which currently looks like:
```
set _dbt_max_partition = (
    select max(<partition_column>) from `<projectid>`.`<dataset>`.`<table>`
);
```
could become something more like:
```
set _dbt_max_partition = (
  select max(parse_date('%Y%m%d', partition_id))
  from `<projectid>`.`<dataset>`.INFORMATION_SCHEMA.PARTITIONS
  where table_name = '<table>'
);
```
for date partitions; with the appropriate handling of other supported partition types and granularity.
